File: std/php/microtime.go
package php

import (
	"fmt"
	"time"

	"github.com/php-any/origami/data"
	"github.com/php-any/origami/node"
)

func NewMicrotimeFunction() data.FuncStmt {
	return &MicrotimeFunction{}
}

type MicrotimeFunction struct {
	data.Function
}

func (f *MicrotimeFunction) Call(ctx data.Context) (data.GetValue, data.Control) {
	now := time.Now()

	// PHP microtime(false): sprintf("%.8F %ld", usec/1e6, sec)  → "0.12345600 1699999999"
	// 旧实现写成 "%d %d"（微秒整数 秒），CombGenerator 的 substr($time[0], 2, 5)
	// 会拿到空串，同一秒内时间戳部分完全相同。
	getAsFloat := false
	temp, ok := ctx.GetIndexValue(0)
	if ok {
		if ab, ok := temp.(data.AsBool); ok {
			getAsFloat, _ = ab.AsBool()
		}
	}
	if getAsFloat {
		microseconds := float64(now.Unix()) + float64(now.Nanosecond())/1e9
		return data.NewFloatValue(microseconds), nil
	}
	sec := now.Unix()
	frac := float64(now.Nanosecond()) / 1e9
	result := fmt.Sprintf("%.8f %d", frac, sec)
	return data.NewStringValue(result), nil
}

func (f *MicrotimeFunction) GetName() string {
	return "microtime"
}

func (f *MicrotimeFunction) GetParams() []data.GetValue {
	return []data.GetValue{
		node.NewParameter(nil, "get_as_float", 0, data.NewBoolValue(false), nil),
	}
}

func (f *MicrotimeFunction) GetVariables() []data.Variable {
	return []data.Variable{
		node.NewVariable(nil, "get_as_float", 0, nil),
	}
}

File: tests/php/microtime.php
<?php

echo "=== microtime() 函数测试 ===\n";

// 测试默认调用（返回字符串格式）
$result = microtime();
if(gettype($result) == "string") {
    Log::info("microtime() 返回字符串格式测试通过");
} else {
    Log::fatal("microtime() 返回字符串格式测试失败，类型: " . gettype($result));
}

// 检查字符串格式（应该是 "微秒 秒" 格式）
if($result->split(" ")->length == 2) {
    Log::info("microtime() 字符串格式正确测试通过");
} else {
    Log::fatal("microtime() 字符串格式正确测试失败，格式: {$result}");
}

// 测试 get_as_float = false（返回字符串）
$result = microtime(false);
if(gettype($result) == "string") {
    Log::info("microtime(false) 返回字符串测试通过");
} else {
    Log::fatal("microtime(false) 返回字符串测试失败，类型: " . gettype($result));
}

// 测试 get_as_float = true（返回浮点数）
$result = microtime(true);
if(gettype($result) == "float") {
    Log::info("microtime(true) 返回浮点数测试通过");
} else {
    Log::fatal("microtime(true) 返回浮点数测试失败，类型: " . gettype($result));
}

// 测试浮点数格式的值应该大于当前 Unix 时间戳
$floatResult = microtime(true);
$currentTime = time();
if($floatResult > $currentTime && $floatResult < $currentTime + 1) {
    Log::info("microtime(true) 返回值在合理范围内测试通过");
} else {
    Log::fatal("microtime(true) 返回值在合理范围内测试失败，值: {$floatResult}, 当前时间: {$currentTime}");
}

// 测试两次调用 microtime(true) 的值应该是递增的
$time1 = microtime(true);
$time2 = microtime(true);
if($time2 >= $time1) {
    Log::info("microtime(true) 时间递增测试通过");
} else {
    Log::fatal("microtime(true) 时间递增测试失败，time1: {$time1}, time2: {$time2}");
}

echo "=== microtime() 测试完成 ===\n";

